src/Type/EntityStorage/EntityStorageDynamicReturnTypeExtension.php

```diff
@@ -100,6 +88,35 @@ public function getTypeFromMethodCall(
             return $returnType;
         }
 
-        return $type;
+        if (!$callerType instanceof EntityStorageType) {
+            $resolvedEntityType = $this->entityDataRepository->resolveFromStorage($callerType);
+            if ($resolvedEntityType === null) {
+                return ParametersAcceptorSelector::selectSingle($methodReflection->getVariants())->getReturnType();
+            }
+            $type = $resolvedEntityType->getClassType();
+        } else {
+            $type = $this->entityDataRepository->get($callerType->getEntityTypeId())->getClassType();
+        }
```

Comment:
This is the new trick for resolving the entity type. This should stay in this PR, I believe as it is specifically used for resolving from `/** @var */`.